lib/hexpm/oauth_providers/github.ex

```diff
@@ -0,0 +1,119 @@
+defmodule Hexpm.OAuthProviders.GitHub do
+  @moduledoc """
+  Provides an interface for the GitHub OAuth integration
+  """
+  use Hexpm.Context
+
+  require Logger
+
+  @base_github_access_token_uri "https://github.com/login/oauth/access_token"
+
+  def authorize_uri do
+    config = Application.get_env(:hexpm, __MODULE__)
+
+    %URI{
+      host: "github.com",
+      scheme: "https",
+      path: "/login/oauth/authorize",
+      query:
+        URI.encode_query(%{
+          "client_id" => config[:client_id],
+          "redirect_uri" => config[:redirect_uri],
+          "scope" => config[:scope]
+        })
+    }
+    |> URI.to_string()
+  end
+
+  @doc """
+  Queries user id from an access_token
+  """
+  def get_user(token) do
+    with {:ok, 200, _headers, body} <-
+           Hexpm.HTTP.get("https://api.github.com/user", headers(token)),
+         {:ok, decoded_body} <- Jason.decode(body),
+         {:ok, user_id} <- Map.fetch(decoded_body, "id"),
+         {:ok, email} <- Map.fetch(decoded_body, "email") do
+      {:ok, %{id: user_id, email: email}}
+    else
+      {:ok, status, _headers, body} ->
+        Logger.error(
+          "[#{__MODULE__}] get_user_id failed with status code: #{inspect(status)} and body #{
+            inspect(body)
+          }"
+        )
+
+      {:error, %Jason.DecodeError{}} = error ->
+        Logger.error(
+          "[#{__MODULE__}] get_user_id failed decoding response with error: #{inspect(error)}"
+        )
+
+        error
+
+      error ->
+        Logger.error(
+          "[#{__MODULE__}] get_user_id failed with unexpected error: #{inspect(error)}"
+        )
```

Comment:
I am usually not a fan of logging errors this way. I think it would be better to just let it crash so it gets reported to rollbar and we get notified.

Reply:
I ain't sure about what you mean. I changed it a bit, so if it still isn't what you mean, feel free to correct me :)

Reply:
@v0idpwn in this case, the error will end up in the logs, which is fine if you look at the logs, but we generally dont. If you do not catch the errors with the `else`, they will end up in Rollbar, which we do check(and have alerts for.)